libs/core/langchain_core/schema/document_transformer.py

```diff
@@ -0,0 +1,70 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Sequence
+
+from langchain_core.documents import Document
+
+
+class DocumentTransformerInterface(ABC):
+    """Interface for document transformation systems.
+
+    The default implementation is the BaseDocumentTransformer.
+    A document transformation system takes a sequence of Documents and returns a
+    sequence of transformed Documents.
+
+    Example:
+        .. code-block:: python
+
+            class EmbeddingsRedundantFilter(BaseDocumentTransformer, BaseModel):
+                embeddings: Embeddings
+                similarity_fn: Callable = cosine_similarity
+                similarity_threshold: float = 0.95
+
+                class Config:
+                    arbitrary_types_allowed = True
+
+                def transform_documents(
+                    self, documents: Sequence[Document], **kwargs: Any
+                ) -> Sequence[Document]:
+                    stateful_documents = get_stateful_documents(documents)
+                    embedded_documents = _get_embeddings_from_stateful_docs(
+                        self.embeddings, stateful_documents
+                    )
+                    included_idxs = _filter_similar_embeddings(
+                        embedded_documents, self.similarity_fn, self.similarity_threshold
+                    )
+                    return [stateful_documents[i] for i in sorted(included_idxs)]
+
+                async def atransform_documents(
+                    self, documents: Sequence[Document], **kwargs: Any
+                ) -> Sequence[Document]:
+                    raise NotImplementedError
+
+    """  # noqa: E501
+
+    @abstractmethod
+    def transform_documents(
+        self, documents: Sequence[Document], **kwargs: Any
+    ) -> Sequence[Document]:
```

Comment:
Return type should be concrete and ownership over list belongs to the user -- we need to fix this across the code base

```suggestion
    ) -> List[Document]:
```

Reply:
> Return type should be concrete and ownership over list belongs to the user -- we need to fix this across the code base

Not sure, why return type should be concrete.
I've found these arguments:
In favor of interfaces: "Whether I own the calling code makes a difference too (internal vs public APIs):

If I own the code that calls the method in question (i.e. internal API), then the more willing I am to return the concrete type.
If I don't control the code that calls this method (e.g. public API), the more likely I am to return the interface type instead. Returning the concrete type is a commitment and, generally speaking, the less I promise, the easier.
Other considerations:

Testing may be easier with interfaces since I can use a mock object that implements the interface..."

In favor of concrete: 
"Rule of thumb, in return types, be a as specific as possible, in parameter types be as unspecific as possible." - which looks like dogma. 

Anyway, I don't get it, why?